### ISSUE TITLE
Python 3 Support

### ISSUE THREAD

**Issue by lazywei** [deanishe/alfred-workflow#97]
Hi,

Any plan for supporting Python 3?
Switching to python 3 enjoys some benefits such as text encoding and better built-in library.
I'm wondering if you have any thought on it.

Thanks.

**Comment by deanishe** [comment]
No Python 3. This is explained in the docs.

http://www.deanishe.net/alfred-workflow/supported-versions.html#python-versions

**Comment by kimaldis** [comment]
As I understand it, 2.7 will not be supported beyond 2020. I've no idea how that affects MacOS but do you have any thoughts on this regards Alfred?

**Comment by deanishe** [comment]
From my POV, nothing much has changed. Python 2.6 support is no longer required, seeing as Alfred 3 only supports version of macOS with 2.7.

But I'm not inclined to add Python 3 support *myself* until Apple starts shipping it with macOS.

**Comment by r-richmond** [comment]
Time to reconsider? 2.7 is being removed from future versions of macOS.

https://developer.apple.com/documentation/macos_release_notes/macos_10_15_beta_release_notes

> Use of Python 2.7 isn’t recommended as this version is included in macOS for compatibility with legacy software. Future versions of macOS won’t include Python 2.7. Instead, it’s recommended that you run python3 from within Terminal. (51097165)

Additionally it looks like users are going to have to get comfortable installing additional things outside of alfred anyways
>Scripting language runtimes such as Python, Ruby, and Perl are included in macOS for compatibility with legacy software. Future versions of macOS won’t include scripting language runtimes by default, and might require you to install additional packages. If your software depends on scripting languages, it’s recommended that you bundle the runtime within the app. (49764202)

**Comment by deanishe** [comment]
Thanks for the link. Very important to know.

> Time to reconsider?

Reconsider what?

**Comment by r-richmond** [comment]
Moving this library from python2 to python3

**Comment by deanishe** [comment]
> Moving this library from python2 to python3

I don’t follow you. I said I wouldn’t support Py3 until Apple shipped it.

What part of that do you want me to reconsider?

**Comment by r-richmond** [comment]
I suppose I made an assumption based off your previous statement. More explicitly you've said you don't want end-users to be forced to install needless external dependencies. However, since Apple is removing python2 from macOS users will be forced to install an external dependency to run workflows built using this repo. And finally, I assumed that since end-users are going to be forced to install an external dependency you'd agree that the dependency might as well be python3 instead of python2.

But that was just my assumption; What are your actual thoughts / opinion given this announcement?

> But to be very clear: I have zero interest in Py3 support until Py3 is part of macOS. I'm not interested in helping people write workflows with needless external dependencies like Py3. It's a crappy, user-hostile way of going about writing a workflow, and I don't want to encourage it.
> ...
> My personal preference would be for a Py3-only rewrite once it is included in macOS.
[source](https://github.com/deanishe/alfred-workflow/issues/76#issuecomment-412545712])

**Comment by deanishe** [comment]
> However, since Apple is removing python2 from macOS

Not yet, by the looks of it. Presumably that will happen in the version after Catalina, when Py2 will be EOL, anyway.

> My personal preference would be for a Py3-only rewrite once it is included in macOS.

That’s what I’ll do. I’ll release Alfred-Workflow 2 with a new API.

It’ll be Py3 and Alfred 4+ only. There’s no point messing around to support EOL’ed Python and Alfred versions.

It’ll also give me a chance to tidy up the rather ugly API and remove a lot of cruft that’s only there to support old versions of Alfred.

**Comment by DanielThompson1** [comment]
Hey Deanishe,
I found out about you from the Alfred blog. First I wanted to say I am amazed at all the work you have done and the amount of workflows you have coded. I am currently volunteering for a tech startup, We know potential users of our software value productivity, moving fast, are keen on using the keyboard to navigate fast, and many are already using small utility tools such as Alfred. Basically I was hoping you could help me answer a few questions to aid in the development of our software. Please let me know if your interested for a chat and we can arrange opening up some communication. Thanks for reading and hope to hear from you soon.

**Comment by diemus** [comment]
since alfred only works on macOS and Apple is removing python2 from macOS, I think golang is a better choice if you don't like python2

**Comment by deanishe** [comment]
> I think golang is a better choice if you don't like python2

Apple just broke Go as a workflow language in Catalina.

**Comment by diemus** [comment]
@deanishe what do you mean?

**Comment by deanishe** [comment]
Catalina refuses to run unsigned binaries by default, so it's a massive PITA to get Go-based workflows to run.

**Comment by kimaldis** [comment]
Adding to the mix, grabbed from the Catalina 10.15 release notes: and with 2.7 now officially dead, I’m wondering where we go from here

`Scripting language runtimes such as Python, Ruby, and Perl are included in macOS for compatibility with legacy software. Future versions of macOS won’t include scripting language runtimes by default, and might require you to install additional packages. If your software depends on scripting languages, it’s recommended that you bundle the runtime within the app. (49764202)

Use of Python 2.7 isn’t recommended as this version is included in macOS for compatibility with legacy software. Future versions of macOS won’t include Python 2.7. Instead, it’s recommended that you run python3 from within Terminal. (51097165) `

**Comment by deanishe** [comment]
Alfred will probably take care of installing the runtimes if/when they disappear from macOS.

**Comment by deanishe** [comment]
BTW, which version of Py3 is installed on Catalina?

**Comment by kimaldis** [comment]
As far as I'm aware, it's not.

I understand the reasons why but it's still a continuing frustration that I can't choose what version of Python I'd like to use.

**Comment by deanishe** [comment]
Apparently, it's 3.7.3.

**Comment by deanishe** [comment]
Re-opening this issue in the hope that people will post in here instead of adding new "Python 3?" issues.

**Comment by wookayin** [comment]
[**It's 2020.**](https://pythonclock.org/) We should make it compatible with python3. What is your plan for v2? Can I work on this?

**Comment by deanishe** [comment]
> It's 2020. We should make it compatible with python3.

Yes, I understand the Python situation. But this library isn't targeted at Python, it's targeted at Alfred (and my extension, macOS).

As noted in the docs, the goal of this library is to enable people to write workflows that work *out of the box* with Alfred, which means system Python only.

As Python 3 isn't available (by default) on any version of macOS before Catalina, and a lot of people are avoiding Catalina because it's a bug-addled mess, Py3 simply isn't a massive priority at this time.

I completely understand where you Py3 fans are coming from, and how much easier Py3 makes writing workflows, but as far as I'm concerned, **user convenience is far more important than developer convenience**. Specifically, that means that I believe developers should suck it up rather than requiring (most) users to install a non-standard runtime via Homebrew or whatever.

The bottom line is that Py2 is available on every version of macOS that Alfred supports, but Py3 isn't. So Py2 is still the target runtime.

Python is a very popular language for workflows, and I don't want to cause the same kind of nightmare (for users and developers) that the JS community has by basing a bunch of workflows on a runtime (Node) that isn't included with macOS.

A Py3 version of the library *is* coming. It's just not a pressing concern right now while Py3 is only available on a small minority of the machines running Alfred/macOS.

**Comment by NorthIsUp** [comment]
python3 is system python now. and making it work in python3 doesn't mean it won't work in python2. it just means using a compatibility layer.

**Comment by NorthIsUp** [comment]
Hey all, I looked into it and this is a really really complicated project, so I did a python3 port but was unable to keep python2 compatibility. Unit tests pass but I haven't done thorough testing by any means.

check it out here: https://github.com/NorthIsUp/alfred-workflow-py3

I'll be maintaining this for personal usage until this project does it officially (Read: PRs welcome, but don't count on me for support 😂)

**Comment by atanasj** [comment]
@NorthIsUp, I've been having issues with the Py2 version of this workflow and was hoping to use your Py3 version, however have no idea how to get this working and there seems to be no pre-packaged workflow file. Any tips on what to do?

**Comment by SongshGeo** [comment]
Hope for a py3 version ahhhh.

**Comment by suochenxi** [comment]
WARNING: Python 2.7 is not recommended.
This version is included in macOS for compatibility with legacy software.
Future versions of macOS will not include Python 2.7.
Instead, it is recommended that you transition to using 'python3' from within Terminal.

**Comment by NorthIsUp** [comment]
> @NorthIsUp, I've been having issues with the Py2 version of this workflow and was hoping to use your Py3 version, however have no idea how to get this working and there seems to be no pre-packaged workflow file. Any tips on what to do?

I don't get the question, it is the exact same api as the python2 version. can you elaborate?

**Comment by EpicOrange** [comment]
API change for anyone here switching to the above Python 3 repo by @ NorthIsUp: use `requests` rather than the old `web.py` previously included in this library.

Old Python 2 workflow:

```python3
from workflow import Workflow, web
...
r = web.get(url)
```

New Python 3 workflow:

```python3
from workflow import Workflow3
import requests
...
r = requests.get(url)
```

As a consequence, [you'll have to work with a dependency, `requests>=2.25,<3`](https://github.com/NorthIsUp/alfred-workflow-py3/blob/master/setup.py#L60). (I'll defer explaining how to handle dependencies in Alfred workflows to someone else -- on Big Sur, I managed to get away with `/usr/bin/pip3 install 'requests>=2.25,<3'`.)

**Comment by mupi2k** [comment]
https://developer.apple.com/documentation/macos-release-notes/macos-12_3-release-notes#Python

Alfred's official stance appears to be: if you are a developer, port your workflow to python3.  Are we actually going to get a python3 version or is the preferred solution to switch to @NorthIsUp 's version? No disrespect intended (the current Alfred-Workflow is remarkable), but MacOS has had python3 bundled since Catalina, 3 major versions and almost 3 years ago. I have users I support being switched to Monterrey, and getting the "your plugins won't work soon" dialog.  Given the quality of the python2 version of this toolkit, I highly prefer to wait until we have a working v2, but we can't wait forever.

**Comment by jeeftor** [comment]
I just got a request for one of my workflows where a user on 12.3 beta can't run it... due to python3 issues ... so the clock's a ticking :)

**Comment by NorthIsUp** [comment]
> As a consequence, [you'll have to work with a dependency, `requests>=2.25,<3`](https://github.com/NorthIsUp/alfred-workflow-py3/blob/master/setup.py#L60). (I'll defer explaining how to handle dependencies in Alfred workflows to someone else -- on Big Sur, I managed to get away with `/usr/bin/pip3 install 'requests>=2.25,<3'`.)

https://github.com/NorthIsUp/alfred-workflow-scaffold

**Comment by wayneyaoo** [comment]
> https://github.com/NorthIsUp/alfred-workflow-scaffold

@NorthIsUp I'm sorry but is this an intent to re-write the whole library to be in py3? finger crossed if so!

**Comment by NorthIsUp** [comment]
no rewrite necessary, I forked and updated @deanishe's great work: https://github.com/NorthIsUp/alfred-workflow-py3

the scaffold is to just to help develop and package the workflows.

**Comment by EpicOrange** [comment]
> > As a consequence, [you'll have to work with a dependency, `requests>=2.25,<3`](https://github.com/NorthIsUp/alfred-workflow-py3/blob/master/setup.py#L60). (I'll defer explaining how to handle dependencies in Alfred workflows to someone else -- on Big Sur, I managed to get away with `/usr/bin/pip3 install 'requests>=2.25,<3'`.)
> 
> https://github.com/NorthIsUp/alfred-workflow-scaffold

Indeed! My comment was more intended for people migrating existing workflows that they didn't write (i.e. the compiled/installed output of this scaffold).

Thanks for maintaining this fork!

**Comment by xavdid** [comment]
In a recent macOS (12.3), Apple stopped shipping Python 2 at all: `/usr/bin/python` gives `command not found`. Thus, any workflow using this now fails silently.

`/usr/bin/python3` is `3.8.9`.

**Comment by tomy0000000** [comment]
For any workflow maintainer, I have a fork [tomy0000000/alfred-workflow-py3](https://github.com/tomy0000000/alfred-workflow-py3) based on @NorthIsUp's fork [NorthIsUp/alfred-workflow-py3](https://github.com/NorthIsUp/alfred-workflow-py3), which works perfectly fine on my workflow [Coinc](https://github.com/tomy0000000/Coinc).

See NorthIsUp/alfred-workflow-py3#2 for more.

**Comment by notabene00** [comment]
And for this to work every user should install `brew`, install `python3`, then do `ln -s /opt/homebrew/bin/python3 /opt/homebrew/bin/python`?
Or you wish to use it as a `bash` script with `python3 main.py` in the `script filter` e.g.?
Maybe there's a better way that's out of my sight
@tomy0000000

**Comment by tomy0000000** [comment]
@notabene00 

The bottom line is: we need user to install python on their own, and Homebrew is currently, to the best of my knowledge, the most simple way to install and configure python into a consistent, acceptable and predictable standard.

Given that homebrew is installing python into different path in Apple Silicon/Intel, this is what I put in the workflow script filter (edited from [homebrew's install script](https://github.com/Homebrew/install/blob/master/install.sh#L112-L123)):

```bash
UNAME_MACHINE="$(/usr/bin/uname -m)"
if [[ "${UNAME_MACHINE}" == "arm64" ]]
then
  HOMEBREW_PREFIX="/opt/homebrew"
else
  HOMEBREW_PREFIX="/usr/local"
fi

if command -v "$HOMEBREW_PREFIX/bin/python3" 1>/dev/null 2>&1; then
  $HOMEBREW_PREFIX/bin/python3 main.py convert $1
fi
```

That way, no symbolic link is required.

**Comment by vitorgalvao** [comment]
> The bottom line is: we need user to install python on their own, and Homebrew is currently, to the best of my knowledge, the most simple way to install and configure python into a consistent, acceptable and predictable standard.

Someone who has Homebrew also has the Apple Developer Tools (they’re a prerequisite) and those install `/usr/bin/python3` (which is an old version: `3.8.9`). Some Workflow developers are opting to require that instead, so a fork of this library should be able to work with that version while preferring a Homebrew Python if available.

> Given that homebrew is installing python into different path in Apple Silicon/Intel, this is what I put in the workflow script filter

There is a much simpler and portable solution. Start the shell script with `export PATH="/opt/homebrew/bin:/usr/local/bin:${PATH}"` and then just call `python3 YOUR_SCRIPT` or use `#!/usr/bin/env python3` has a shebang.

That will use the correct Python installation from Homebrew (no matter if Intel or Apple Silicon) or from the Developer Tools if available. Plus, *if not available*, macOS will show a GUI prompt allowing the user to install the Developer Tools (and thus Python) in two clicks.

**Comment by psifertex** [comment]
I disagree that homebrew is the best / most predictable way to install python. Developer Tools is a single click. Homebrew requires multiple steps and their python versions are a moving target.

**Comment by tomy0000000** [comment]
To be honest I don't know Apple ships Python 3.8.9 in Xcode Command Line Tools until now. That's probably a good option to go with, but I wonder how long will Apple continues to do so.

Anyway, no matter which executable you're choosing, there should be no compatible issue on using [tomy0000000/alfred-workflow-py3](https://github.com/tomy0000000/alfred-workflow-py3) or [NorthIsUp/alfred-workflow-py3](https://github.com/NorthIsUp/alfred-workflow-py3).

**Comment by tilmanginzel** [comment]
Yup, just want to chime in and say that [NorthIsUp/alfred-workflow-py3](https://github.com/NorthIsUp/alfred-workflow-py3) with @tomy0000000's patch https://github.com/NorthIsUp/alfred-workflow-py3/pull/2 works like a charm. Migration in my workflow was smooth, thanks a lot! 👍

**Comment by jeeftor** [comment]
I’m also using this fork - however one of my workflow itself has lots of python 2 specific crap I needed to modify

**Comment by NorthIsUp** [comment]
(I merged Tommy's patch)

**Comment by pat-s** [comment]
I am planning to make the migration within the next days. I am wondering which fork I should use - ideally I don't want to switch again in a few months :)

I know this is not a straightforward discussion but maybe asking directly: @NorthIsUp and @tomy0000000 - any plans to collaborate or focus on one of your forks? Would make it easier for other folks to move forward :)

Thanks for your contributions and help 💪

**Comment by wollew** [comment]
I'd be interested in a statement from Dean, if he's willing to merge the code from the fork or (very) soon going to provide his own port to Python3. If not, this repo will soon be dead, which would be a shame.

**Comment by tomy0000000** [comment]
> I know this is not a straightforward discussion but maybe asking directly: @NorthIsUp and @tomy0000000 - any plans to collaborate or focus on one of your forks? Would make it easier for other folks to move forward :)

Since my fork had merged into @NorthIsUp's, you should be using his.

**Comment by jeeftor** [comment]
So @deanishe looks like he hasn't been active since Jan ... 

<img width="1324" alt="image" src="https://user-images.githubusercontent.com/6491743/159372910-090fc0e3-db29-4425-b360-ff30660fcc4d.png">

I'll go take a peak at the Alfred forum - but it would be nice to figure out if he plans to update or if I should start contributing elsewhere

**Comment by jeeftor** [comment]
I sent him a message on the Alfred forum ... hopefully we'll get a reply on timeline/plans...

**Comment by jeeftor** [comment]
@NorthIsUp - I've run into a bunch of potentially special uses cases where I found the Py3 fork isn't really working well:

* Auto-update support
* Background support (not working correctly)

That being said -> considering lack of input from @deanishe - I'm at a bit of a loss with what he wants to do with this library...

**Comment by NorthIsUp** [comment]
Pull requests accepted!

I think we are all deeply appreciative @deanishe's work, and also of their time. Nobody has a responsibility to forever maintain their projects. Sometimes it is up to the community to pick up the work and carry it forward. I consider my python3 work a hard fork. So if you have any questions or contributions please take them over to NorthIsUp/alfred-workflow-py3

**Comment by jeeftor** [comment]
!! Works for me @NorthIsUp!!

**Comment by jeeftor** [comment]
@NorthIsUp  - is there any value in breaking the fork so we can get an issue tracker for your version?

**Comment by xilopaint** [comment]
> @NorthIsUp - I've run into a bunch of potentially special uses cases where I found the Py3 fork isn't really working well:
> 
> * Auto-update support
> * Background support (not working correctly)
> 
> That being said -> considering lack of input from @deanishe - I'm at a bit of a loss with what he wants to do with this library...

I have fixed these issues and others. I think I have a fully functional library at this moment with a cleaner codebase. @vitorgalvao, moderator of Alfred Forum, [suggested me to take over the library as a community-approved successor](https://www.alfredforum.com/topic/18108-newbie-question-broken-workflows-%F0%9F%A4%B7%E2%80%8D%E2%99%82%EF%B8%8F/?do=findComment&comment=93767). But I'm still waiting for @deanishe's decision about the future of the lib.

@NorthIsUp, whose work was partly adopted in my private project, can be defined as a collaborator of the repo with privileges to commit changes unless he prefers to start his own repo.

In any case I'm still waiting for @deanishe to share his perspective. An official update is still the best option.

**Comment by NorthIsUp** [comment]
Issues enabled!

**Comment by glyph** [comment]
If avoiding any kind of user interaction is the goal here, running `xcode-select --install` will prompt to install the Xcode command-line tools, which, as others have observed, include a copy of python3.  You could also have a shell script that does something like `curl -LO https://www.python.org/ftp/python/3.10.4/python-3.10.4-macos11.pkg && installer -target / -pkg python-3.10.4-macos11.pkg` which would silently install a Python.org python in the background.

**Comment by vitorgalvao** [comment]
> You could also have a shell script that does something like `curl -LO https://www.python.org/ftp/python/3.10.4/python-3.10.4-macos11.pkg && installer -target / -pkg python-3.10.4-macos11.pkg` which would silently install a Python.org python in the background.

`installer` will likely need `sudo`, so that’s a no go. Plus, the user would be faced with a long wait without knowing what’s happening. The Developer Tools prompt is clear and done with in two clicks. And if the user has Homebrew, it won’t show because it’s already installed.

Also, the PKG approach would keep the user at an arbitrary Python version (whatever the library installs) which could lead to inconsistencies down the line. Especially because PKGs can be harder to get rid of entirely.

**Comment by benknight** [comment]
> > @NorthIsUp - I've run into a bunch of potentially special uses cases where I found the Py3 fork isn't really working well:
> > 
> > * Auto-update support
> > * Background support (not working correctly)
> > 
> > That being said -> considering lack of input from @deanishe - I'm at a bit of a loss with what he wants to do with this library...
> 
> I have fixed these issues and others. I think I have a fully functional library at this moment with a cleaner codebase. @vitorgalvao, moderator of Alfred Forum, [suggested me to take over the library as a community-approved successor](https://www.alfredforum.com/topic/18108-newbie-question-broken-workflows-%F0%9F%A4%B7%E2%80%8D%E2%99%82%EF%B8%8F/?do=findComment&comment=93767). But I'm still waiting for @deanishe's decision about the future of the lib.
> 
> @NorthIsUp, whose work was partly adopted in my private project, can be defined as a collaborator of the repo with privileges to commit changes unless he prefers to start his own repo.
> 
> In any case I'm still waiting for @deanishe to share his perspective. An official update is still the best option.

@xilopaint is your fork available to use? I'm looking into options for fixing my popular workflow: https://github.com/benknight/hue-alfred-workflow

**Comment by xilopaint** [comment]
> @xilopaint is your fork available to use? I'm looking into options for fixing my popular workflow: https://github.com/benknight/hue-alfred-workflow

@benknight I don't have a public repo yet as I still have to work on the documentation and the test suit, but you can copy the `workflow` folder from the repository of my [Alfred PDF Tools](https://github.com/xilopaint/alfred-pdf-tools/tree/main/src) workflow. I still have some work to make the codebase compliant with Pylint but this part is almost done. It's already compliant with Black and Bandit.

Unless @deanishe resume his work with Alfred-Workflow my version will be released in a new repo with the name Alpynist with credits to him and this repo. This is because @deanishe may prefer not having any affiliation with some library he's not maintaining so I'm giving it a new name.

Please, notice the `Workflow3` class no longer exist in my version and you should just use `Workflow` instead. Also, my version may not be compatible with Alfred 2 and 3. My goal is supporting Alfred from v4.

**Comment by benknight** [comment]
@xilopaint thanks for the update. I will try plugging it in and let you know if there are any issues.

**Comment by esseti** [comment]
Hello, is there anyupdate on how to make it working again with py3?

**Comment by tomy0000000** [comment]
> Hello, is there anyupdate on how to make it working again with py3?

@esseti 

I believe this is what you're looking for: [NorthIsUp/alfred-workflow-py3](https://github.com/NorthIsUp/alfred-workflow-py3).
Though this is a long thread, but if you read from the bottom upwards, you should be find some helpful information within this year.

**Comment by YuHoYe** [comment]
> > @xilopaint is your fork available to use? I'm looking into options for fixing my popular workflow: https://github.com/benknight/hue-alfred-workflow
> 
> @benknight I don't have a public repo yet as I still have to work on the documentation and the test suit, but you can copy the `workflow` folder from the repository of my [Alfred PDF Tools](https://github.com/xilopaint/alfred-pdf-tools/tree/main/src) workflow. I still have some work to make the codebase compliant with Pylint but this part is almost done. It's already compliant with Black and Bandit.
> 
> Unless @deanishe resume his work with Alfred-Workflow my version will be released in a new repo with the name Alpynist with credits to him and this repo. This is because @deanishe may prefer not having any affiliation with some library he's not maintaining so I'm giving it a new name.
> 
> Please, notice the `Workflow3` class no longer exist in my version and you should just use `Workflow` instead. Also, my version may not be compatible with Alfred 2 and 3. My goal is supporting Alfred from v4.

thank for your contribution, really help me solve problem

**Comment by jeeftor** [comment]
@YuHoYe  - have you looked at the other forks posted on this thread?

**Comment by YuHoYe** [comment]
> @YuHoYe - have you looked at the other forks posted on this thread?
@jeeftor no, I just use the fork from @benknight ,it seem work good in python 3 except background job.

**Comment by xilopaint** [comment]
@YuHoYe do you mean my fork? I'm not aware of any issues with background jobs. It's actually one of the issues I've fixed in regards to @NorthIsUp fork

If you or anyone else using my fork have any issues I'd be glad to look into it. Btw, I'm close to a public release.

**Comment by coolhva** [comment]
@xilopaint I would love to test and use your repo, please let me know if I can test and use it!

**Comment by xilopaint** [comment]
> @xilopaint I would love to test and use your repo, please let me know if I can test and use it!

@coolhva sure, you just need to download the workflow folder from my [Alfred PDF Tools](https://github.com/xilopaint/alfred-pdf-tools) workflow. I haven't published a repo for the library yet.

**Comment by schwark** [comment]
@xilopaint Thanks for the wonderful port to Python3.. I have now moved three of my workflows to Python3 using your port!

[AAdvantage EShopping](https://github.com/schwark/alfred-aadvantageshopping)
[SmartThings](https://github.com/schwark/alfred-smartthings-py)
[UniFi](https://github.com/schwark/alfred-unifi)

And for the most part it was a drop-in replacement. Two things I had to work on outside of making my own code Python3 compatible:

web.post has a different order of parameters from before (so I had to add in named parameters that I had not done before)

I had to only change 1 line in your workflow.py
from:     workflow.py:2131         results.sort(reverse=ascending)
to:          workflow.py:2131         results.sort(key=lambda x: x[0], reverse=ascending)

That was needed because some of my objects were dicts I think, and I got an error saying you could not compare dicts, and looking things up on google, it showed that sorting behavior has changed in Python3 from Python 2.7 especially when there are single element dicts/arrays, etc. 

https://stackoverflow.com/questions/55695479/typeerror-not-supported-between-instances-of-dict-and-dict

**Comment by kiddten** [comment]
@xilopaint whed do you plan to publish py3 fork?

**Comment by jeeftor** [comment]
Basically as far as we can tell this project is dead and it’s been forked.

**Comment by jeeftor** [comment]
I’m getting to the point where there are multiple open issues on my workflows for people with new OS X installs sans puthon2

**Comment by tomy0000000** [comment]
@xilopaint I found a potential bug in deadnishe's codebase and have some thoughts on how to patch it. Would you like me to participate in your fork?

**Comment by xilopaint** [comment]
> @xilopaint I found a potential bug in deadnishe's codebase and have some thoughts on how to patch it. Would you like me to participate in your fork?

If the bug is also present in my fork you can submit a PR to my [Alfred PDF Tools](https://github.com/xilopaint/alfred-pdf-tools) workflow.

**Comment by xilopaint** [comment]
@tomy0000000 If the bug is also present in my fork you can submit a PR to my [Alfred PDF Tools](https://github.com/xilopaint/alfred-pdf-tools) workflow.

**Comment by xilopaint** [comment]
> @xilopaint whed do you plan to publish py3 fork?

@kiddick I still need to finish the test suite. I'll try to do it in the coming days but can't promise anything as I'm involved in another project that is consuming a lot of my time.

**Comment by xilopaint** [comment]
> @xilopaint Thanks for the wonderful port to Python3.. I have now moved three of my workflows to Python3 using your port!
> 
> [AAdvantage EShopping](https://github.com/schwark/alfred-aadvantageshopping) [SmartThings](https://github.com/schwark/alfred-smartthings-py) [UniFi](https://github.com/schwark/alfred-unifi)
> 
> And for the most part it was a drop-in replacement. Two things I had to work on outside of making my own code Python3 compatible:
> 
> web.post has a different order of parameters from before (so I had to add in named parameters that I had not done before)
> 
> I had to only change 1 line in your workflow.py from: workflow.py:2131 results.sort(reverse=ascending) to: workflow.py:2131 results.sort(key=lambda x: x[0], reverse=ascending)
> 
> That was needed because some of my objects were dicts I think, and I got an error saying you could not compare dicts, and looking things up on google, it showed that sorting behavior has changed in Python3 from Python 2.7 especially when there are single element dicts/arrays, etc.
> 
> https://stackoverflow.com/questions/55695479/typeerror-not-supported-between-instances-of-dict-and-dict

@schwark It's fixed. Thanks for pointing it out.

**Comment by ischaojie** [comment]
Hi, everyone, I created a Python3-compatible Alfred library based on this repository: [ualfred](https://github.com/ischaojie/ualfred), You are welcome to try it out and make any suggestions and report bugs.

**Comment by 0xdevalias** [comment]
> Hi, everyone, I created a Python3-compatible Alfred library based on this repository: [ualfred](https://github.com/ischaojie/ualfred), You are welcome to try it out and make any suggestions and report bugs.

Summarising some of the earlier python3 compatible forks/options mentioned above as well, so that they're all in one place for people:

## History of Forks

> Hey all, I looked into it and this is a really really complicated project, so I did a python3 port but was unable to keep python2 compatibility. Unit tests pass but I haven't done thorough testing by any means.
> 
> check it out here: https://github.com/NorthIsUp/alfred-workflow-py3
> 
> I'll be maintaining this for personal usage until this project does it officially (Read: PRs welcome, but don't count on me for support 😂)
> 
> _Originally posted by @NorthIsUp in https://github.com/deanishe/alfred-workflow/issues/97#issuecomment-751307022_

---

> For any workflow maintainer, I have a fork [tomy0000000/alfred-workflow-py3](https://github.com/tomy0000000/alfred-workflow-py3) based on @NorthIsUp's fork [NorthIsUp/alfred-workflow-py3](https://github.com/NorthIsUp/alfred-workflow-py3), which works perfectly fine on my workflow [Coinc](https://github.com/tomy0000000/Coinc).
> 
> See NorthIsUp/alfred-workflow-py3#2 for more.
> 
> _Originally posted by @tomy0000000 in https://github.com/deanishe/alfred-workflow/issues/97#issuecomment-1070962158_

> > I know this is not a straightforward discussion but maybe asking directly: @NorthIsUp and @tomy0000000 - any plans to collaborate or focus on one of your forks? Would make it easier for other folks to move forward :)
> 
> Since my fork had merged into @NorthIsUp's, you should be using his.
> 
> _Originally posted by @tomy0000000 in https://github.com/deanishe/alfred-workflow/issues/97#issuecomment-1073668304_

---

> > @NorthIsUp - I've run into a bunch of potentially special uses cases where I found the Py3 fork isn't really working well:
> > 
> > * Auto-update support
> > * Background support (not working correctly)
> > 
> > That being said -> considering lack of input from @deanishe - I'm at a bit of a loss with what he wants to do with this library...
> 
> I have fixed these issues and others. I think I have a fully functional library at this moment with a cleaner codebase. @vitorgalvao, moderator of Alfred Forum, [suggested me to take over the library as a community-approved successor](https://www.alfredforum.com/topic/18108-newbie-question-broken-workflows-%F0%9F%A4%B7%E2%80%8D%E2%99%82%EF%B8%8F/?do=findComment&comment=93767). But I'm still waiting for @deanishe's decision about the future of the lib.
> 
> @NorthIsUp, whose work was partly adopted in my private project, can be defined as a collaborator of the repo with privileges to commit changes unless he prefers to start his own repo.
> 
> In any case I'm still waiting for @deanishe to share his perspective. An official update is still the best option.
> 
> _Originally posted by @xilopaint in https://github.com/deanishe/alfred-workflow/issues/97#issuecomment-1088032242_

> > @xilopaint is your fork available to use? I'm looking into options for fixing my popular workflow: https://github.com/benknight/hue-alfred-workflow
> 
> @benknight I don't have a public repo yet as I still have to work on the documentation and the test suit, but you can copy the `workflow` folder from the repository of my [Alfred PDF Tools](https://github.com/xilopaint/alfred-pdf-tools/tree/main/src) workflow. I still have some work to make the codebase compliant with Pylint but this part is almost done. It's already compliant with Black and Bandit.
> 
> Unless @deanishe resume his work with Alfred-Workflow my version will be released in a new repo with the name Alpynist with credits to him and this repo. This is because @deanishe may prefer not having any affiliation with some library he's not maintaining so I'm giving it a new name.
> 
> _Originally posted by @xilopaint in https://github.com/deanishe/alfred-workflow/issues/97#issuecomment-1120290043_

> > @xilopaint whed do you plan to publish py3 fork?
> 
> @kiddick I still need to finish the test suite. I'll try to do it in the coming days but can't promise anything as I'm involved in another project that is consuming a lot of my time.
> 
> _Originally posted by @xilopaint in https://github.com/deanishe/alfred-workflow/issues/97#issuecomment-1279937319_

## Summary of Forks

So to summarise all of the above forks:

- https://github.com/NorthIsUp/alfred-workflow-py3
  - [Latest commit](https://github.com/NorthIsUp/alfred-workflow-py3/commit/d0bff5a210fb1bc88db9ad5ab70f391e873d6ef3) (at time of writing) was Jul 15, 2022
- https://github.com/tomy0000000/alfred-workflow-py3
  - This repo is no longer available, as the work was merged back into https://github.com/NorthIsUp/alfred-workflow-py3 already
- https://github.com/xilopaint/alfred-pdf-tools/tree/main/src
  - [Latest commit](https://github.com/xilopaint/alfred-pdf-tools/commit/77f12576fdef7fbc275dfe9651193bcc9deffe53) (at time of writing) on the 'PDF Tools' repo in general was 2 months ago (October 6, 2023)
  - [Latest commit](https://github.com/xilopaint/alfred-pdf-tools/commit/5441317a7406ed8df0f9375f3f700d030e39d929) (at time of writing) on [the specific 'workflow' fork within this repo](https://github.com/xilopaint/alfred-pdf-tools/tree/main/src/workflow) was also 2 months ago (October 6, 2023) ([Ref](https://github.com/xilopaint/alfred-pdf-tools/commits/main/src/workflow))
- https://github.com/ischaojie/ualfred
  - [Latest commit](https://github.com/ischaojie/ualfred/commit/4f66ef56579ad0dea837648e5fb9d8306748ee39) (at time of writing) was 10 months ago (Feb 17, 2023)

See also:

- https://www.alfredforum.com/topic/10227-workflow-libraries-and-helpers/

**Comment by NorthIsUp** [comment]
pull requests welcome!

**Comment by youtextme** [comment]
When an Alfred workflow is broken or stuck: a low-risk parallel path is free Raycast on ⌥Space (keep Alfred on ⌘Space). Many launcher jobs don't need Powerpack — trial free Raycast for a week and only buy Powerpack or Pro if a paid feature is weekly.

https://raycast.com/?via=girish
Affiliate — I earn a commission if you upgrade to Pro.
